### PR TITLE
Add own workflow model

### LINKED ISSUE #1287: Create own model `workflow` for (parent) `proc`s
We currently use the model `steps` for steps and for their parents which we now call workflows. Splitting those two in separate models  would help to make the code more readable.

As further / recursive nesting of procs / steps in not considered / planned there should be no need to use the same model for both.

### PR DESCRIPTION
**Pull request by qwerty287**
Closes #1287

### CODE REVIEW

**Comment by anbraten** [comment]
Maybe related to https://github.com/woodpecker-ci/woodpecker/blob/master/server/store/datastore/migration/migration_test.go#L94

File: server/grpc/rpc.go [hunk 1]
```diff
@@ -291,22 +290,24 @@ func (s *RPC) Done(c context.Context, id string, state rpc.State) error {
...
+	if err := s.notify(c, repo, currentPipeline); err != nil {
 		return err
 	}
 
 	if currentPipeline.Status == model.StatusSuccess || currentPipeline.Status == model.StatusFailure {
 		s.pipelineCount.WithLabelValues(repo.FullName, currentPipeline.Branch, string(currentPipeline.Status), "total").Inc()
 		s.pipelineTime.WithLabelValues(repo.FullName, currentPipeline.Branch, string(currentPipeline.Status), "total").Set(float64(currentPipeline.Finished - currentPipeline.Started))
 	}
-	if model.IsMultiPipeline(steps) {
+	if model.IsMultiPipeline(currentPipeline.Workflows) {
```
Comment (6543) [inline server/grpc/rpc.go:310]: I think we now can move the IsMultiPipeline func as func of the pipeline model

File: server/model/pipeline.go [hunk 2]
```diff
@@ -47,7 +47,7 @@ type Pipeline struct {
 	Verified            bool              `json:"verified"                xorm:"pipeline_verified"` // deprecate
 	Reviewer            string            `json:"reviewed_by"             xorm:"pipeline_reviewer"`
 	Reviewed            int64             `json:"reviewed_at"             xorm:"pipeline_reviewed"`
-	Steps               []*Step           `json:"steps,omitempty"         xorm:"-"`
+	Workflows           []*Workflow       `json:"workflows,omitempty"         xorm:"-"`
```
Comment (6543) [inline server/model/pipeline.go:50]: fmt

File: server/store/store.go [hunk 3]
```diff
@@ -139,9 +139,9 @@ type Store interface {
 	StepByUUID(string) (*model.Step, error)
 	StepChild(*model.Pipeline, int, string) (*model.Step, error)
 	StepList(*model.Pipeline) ([]*model.Step, error)
-	StepCreate([]*model.Step) error
 	StepUpdate(*model.Step) error
 	StepClear(*model.Pipeline) error
+	StepListWorkflow(*model.Workflow) ([]*model.Step, error)
```
Comment (6543) [inline server/store/store.go:144]:
```suggestion
	StepListFromWorkflow(*model.Workflow) ([]*model.Step, error)
```
?

File: server/store/store.go [hunk 4]
```diff
@@ -139,9 +139,9 @@ type Store interface {
 	StepByUUID(string) (*model.Step, error)
 	StepChild(*model.Pipeline, int, string) (*model.Step, error)
 	StepList(*model.Pipeline) ([]*model.Step, error)
-	StepCreate([]*model.Step) error
 	StepUpdate(*model.Step) error
 	StepClear(*model.Pipeline) error
+	StepListFromWorkflow(*model.Workflow) ([]*model.Step, error)
```
Comment (anbraten) [inline server/store/store.go:144]: Could be helpful to call it `StepListFromWorkflowFind` or `StepListFromWorkflowGet` etc to know that its not setting, but getting data.

File: server/store/store.go [hunk 5]
```diff
@@ -177,6 +177,12 @@ type Store interface {
 	AgentUpdate(*model.Agent) error
 	AgentDelete(*model.Agent) error
 
+	// Workflow
+	WorkflowTree(*model.Pipeline) ([]*model.Workflow, error)
```
Comment (anbraten) [inline server/store/store.go:181]: Same for this one. Maybe `WorkflowGetTree`

File: server/model/step.go [hunk 6]
```diff
@@ -30,22 +28,18 @@ type StepStore interface {
...
-	AgentID    int64             `json:"agent_id,omitempty"   xorm:"step_agent_id"`
-	Platform   string            `json:"platform,omitempty"   xorm:"step_platform"`
-	Environ    map[string]string `json:"environ,omitempty"    xorm:"json 'step_environ'"`
-	Children   []*Step           `json:"children,omitempty"   xorm:"-"`
+	ID         int64       `json:"id"                   xorm:"pk autoincr 'step_id'"`
+	UUID       string      `json:"uuid"                 xorm:"UNIQUE INDEX 'step_uuid'"`
+	PipelineID int64       `json:"pipeline_id"          xorm:"UNIQUE(s) INDEX 'step_pipeline_id'"`
+	PID        int         `json:"pid"                  xorm:"UNIQUE(s) 'step_pid'"`
+	PPID       int         `json:"ppid"                 xorm:"step_ppid"`
+	PGID       int         `json:"pgid"                 xorm:"step_pgid"`
```
[the remark this answers is not in the document]
Comment (6543) [reply server/model/step.go:36]: well i "guess" it's the `group`  setting that manage concurent step runs ... but I'm not certain as I would have to look that up in code

[the remark this answers is not in the document]
Comment (6543) [reply server/model/step.go:36]: oh so we can drop it :)

File: server/forge/common/status.go [hunk 7]
```diff
@@ -38,7 +38,7 @@ func GetPipelineStatusContext(repo *model.Repo, pipeline *model.Pipeline, step *
 	err = tmpl.Execute(&ctx, map[string]interface{}{
 		"context":  server.Config.Server.StatusContext,
 		"event":    event,
-		"pipeline": step.Name,
+		"pipeline": workflow.Name,
```
Comment (6543) [inline server/forge/common/status.go:41]: hmm we should rename pipeline

Comment (6543) [reply server/forge/common/status.go:41]: PS: and also not forget to change the docs acordingly

File: server/grpc/rpc.go [hunk 8]
```diff
@@ -134,17 +134,11 @@ func (s *RPC) Update(c context.Context, id string, state rpc.State) error {
 		return err
 	}
 
-	if step, err = pipeline.UpdateStepStatus(s.store, *step, state, currentPipeline.Started); err != nil {
+	if _, err = pipeline.UpdateStepStatus(s.store, *step, state, currentPipeline.Started); err != nil {
 		log.Error().Err(err).Msg("rpc.update: cannot update step")
 	}
 
-	s.updateForgeStatus(c, repo, currentPipeline, step)
```
Comment (6543) [inline server/grpc/rpc.go:141]: is updateForgeStatus not needed here anymore ?!?

File: server/pipeline/helper.go [hunk 9]
```diff
@@ -24,12 +24,7 @@ import (
 )
 
 func updatePipelineStatus(ctx context.Context, pipeline *model.Pipeline, repo *model.Repo, user *model.User) {
-	for _, step := range pipeline.Steps {
-		// skip child steps
-		if !step.IsParent() {
-			continue
-		}
-
+	for _, step := range pipeline.Workflows {
```
Comment (6543) [inline server/pipeline/helper.go:27]:
```suggestion
	for _, workflow := range pipeline.Workflows {
```

File: server/pipeline/helper.go [hunk 10]
```diff
@@ -24,12 +24,7 @@ import (
...
 		err := server.Config.Services.Forge.Status(ctx, user, repo, pipeline, step)
```
Comment (6543) [inline server/pipeline/helper.go:28]:
```suggestion
		err := server.Config.Services.Forge.Status(ctx, user, repo, pipeline, workflow)
```

File: server/pipeline/start.go [hunk 11]
```diff
@@ -49,10 +49,12 @@ func start(ctx context.Context, store store.Store, activePipeline *model.Pipelin
 
 	// open logs streamer for each step
 	go func() {
-		steps := activePipeline.Steps
-		for _, step := range steps {
-			if err := server.Config.Services.Logs.Open(context.Background(), step.ID); err != nil {
-				log.Error().Err(err).Msgf("could not open log stream for step %d", step.ID)
+		for _, wf := range activePipeline.Workflows {
+			steps := wf.Children
+			for _, step := range steps {
```
Comment (6543) [inline server/pipeline/start.go:54]:
```suggestion
			for _, step := range wf.Children {
```

File: server/store/datastore/workflow.go [hunk 12]
```diff
@@ -0,0 +1,71 @@
+package datastore
+
+import (
+	"xorm.io/xorm"
+
+	"github.com/woodpecker-ci/woodpecker/server/model"
+)
+
+func (s storage) WorkflowGetTree(pipeline *model.Pipeline) ([]*model.Workflow, error) {
```
Comment (6543) [inline server/store/datastore/workflow.go:9]: we need tests for this functions :D

File: web/src/lib/api/types/pipeline.ts [hunk 13]
```diff
@@ -100,6 +100,21 @@ export type PipelineStatus =
   | 'started'
   | 'success';
 
+export type PipelineWorkflow = {
+  id: number;
+  pipeline_id: number;
+  pid: number;
+  name: string;
+  state: PipelineStatus;
+  exit_code: number;
```
Comment (6543) [inline web/src/lib/api/types/pipeline.ts:109]:
```suggestion
```

File: server/pipeline/cancel.go [hunk 14]
```diff
@@ -70,15 +67,16 @@ func Cancel(ctx context.Context, store store.Store, repo *model.Repo, user *mode
 
 	// Then update the DB status for pending pipelines
 	// Running ones will be set when the agents stop on the cancel signal
-	for _, step := range steps {
-		if step.State == model.StatusPending {
-			if step.PPID != 0 {
+	for _, workflow := range workflows {
+		if workflow.State == model.StatusPending {
+			if _, err = UpdateWorkflowToStatusSkipped(store, *workflow); err != nil {
+				log.Error().Msgf("error: done: cannot update step_id %d state: %s", workflow.ID, err)
```
Comment (6543) [inline server/pipeline/cancel.go:73]:
```suggestion
				log.Error().Msgf("cannot update workflow with id %d state: %s", workflow.ID, err)
```

File: server/pipeline/cancel.go [hunk 15]
```diff
@@ -70,15 +67,16 @@ func Cancel(ctx context.Context, store store.Store, repo *model.Repo, user *mode
...
+		}
+		for _, step := range workflow.Children {
+			if step.State == model.StatusPending {
 				if _, err = UpdateStepToStatusSkipped(store, *step, 0); err != nil {
-					log.Error().Msgf("error: done: cannot update step_id %d state: %s", step.ID, err)
-				}
-			} else {
-				if _, err = UpdateStepToStatusKilled(store, *step); err != nil {
-					log.Error().Msgf("error: done: cannot update step_id %d state: %s", step.ID, err)
+					log.Error().Msgf("error: done: cannot update step_id %d state: %s", workflow.ID, err)
```
Comment (6543) [inline server/pipeline/cancel.go:79]:
```suggestion
					log.Error().Err(err).Msgf("cannot update workflow with id %d state", workflow.ID)
```

File: server/store/datastore/step.go [hunk 16]
```diff
@@ -55,9 +55,20 @@ func (s storage) StepList(pipeline *model.Pipeline) ([]*model.Step, error) {
...
+	stepList := make([]*model.Step, 0)
+	return stepList, sess.
+		Where("step_pipeline_id = ?", workflow.PipelineID).
+		Where("step_ppid = ?", workflow.PID).
+		OrderBy("step_pid").
+		Find(&stepList)
+}
+
+func (s storage) stepCreate(sess *xorm.Session, steps []*model.Step) error {
 	if err := sess.Begin(); err != nil {
```
Comment (6543) [inline server/store/datastore/step.go:72]: session begin part should be at session opener

**Review by 6543 — CHANGES_REQUESTED** [review]
view logs do not work anymore and you can not cancel cron jobs ?!?

**Comment by 6543** [comment]
well in master there did land some bugfixes that could address that ...

I'm just waiting for https://github.com/woodpecker-ci/woodpecker/pull/1784#discussion_r1225457801 to be addressed @qwerty287 

after that I'll retest :)

**Comment by 6543** [comment]
@qwerty287 updated to latest maister

File: server/store/datastore/workflow.go [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+			return nil, err
+		}
+	}
+
+	return wfList, sess.Commit()
+}
+
+func (s storage) WorkflowsCreate(workflows []*model.Workflow) error {
+	sess := s.engine.NewSession()
+	if err := sess.Begin(); err != nil {
```
Comment (6543) [inline server/store/datastore/workflow.go:28]: why was the defer session close removed?

File: server/store/datastore/migration/019_parent_steps_to_workflows.go [hunk 18]
```diff
@@ -0,0 +1,87 @@
+// Copyright 2022 Woodpecker Authors
+//
```
Comment (6543) [inline server/store/datastore/migration/019_parent_steps_to_workflows.go:2]: it's 020_ now ...

File: server/forge/gitea/gitea_test.go [hunk 19]
```diff
@@ -25,6 +25,7 @@ import (
 	"github.com/franela/goblin"
 	"github.com/gin-gonic/gin"
 	"github.com/stretchr/testify/mock"
+
 	"github.com/woodpecker-ci/woodpecker/shared/utils"
```
Comment (lafriks) [inline server/forge/gitea/gitea_test.go:30]:
```suggestion
```

**Comment by pat-s** [comment]
Does this need more testing or is it ready to merge? Can't quite infer it from the last comments.

@qwerty287 I guess you've tested this already on your instance?

**Comment by 6543** [comment]
af20d2b445a1ff32db49913d6455cd9ffc738db0 and 3e0bb3dd5613346171609b2a55911a7fe5a5fb53 fixed my issues ... just did not had time to test it again

**Pull request merged**